engine/notifier_test.go

```diff
@@ -85,123 +87,113 @@ func TestNotifier_ManyNotifications(t *testing.T) {
-// routines consuming work. We require that all worker is eventually processed.
+// routines consuming work. We require that all work is eventually processed.
 func TestNotifier_AllWorkProcessed(t *testing.T) {
 	singleTestRun := func(t *testing.T) {
 		t.Parallel()
-		notifier := NewNotifier()
-
-		totalWork := int32(100)
-		pendingWorkQueue := make(chan struct{}, totalWork)
-		scheduledWork := atomic.NewInt32(0)
-		consumedWork := atomic.NewInt32(0)
-
-		// starts the consuming first, because if we starts the production first instead, then
-		// we might finish pushing all jobs, before any of our consumer has started listening
-		// to the queue.
-		var consumersAllReady sync.WaitGroup
-		consumersAllReady.Add(5)
-
-		// 5 routines consuming work
-		for i := 0; i < 5; i++ {
-			go func() {
-				consumersAllReady.Done()
-				for consumedWork.Load() < totalWork {
-					<-notifier.Channel()
-				L:
+		synctest.Test(t, func(t *testing.T) {
+			ctx, cancel := context.WithCancel(context.Background())
+			defer cancel()
+
+			notifier := NewNotifier()
+
+			totalWork := int32(100)
+			pendingWorkQueue := make(chan struct{}, totalWork)
+			scheduledWork := atomic.NewInt32(0)
+			consumedWork := atomic.NewInt32(0)
+
+			// starts the consumers first, because if we start the production first instead, then
+			// we might finish pushing all jobs, before any of our consumer has started listening
+			// to the queue.
+
+			processAllPending := func() {
+				for {
+					select {
+					case <-pendingWorkQueue:
+						consumedWork.Inc()
+					default:
+						return
+					}
+				}
+			}
+
+			// 5 routines consuming work
+			for range 5 {
+				go func() {
 					for {
 						select {
-						case <-pendingWorkQueue:
-							consumedWork.Inc()
-						default:
-							break L
+						case <-ctx.Done():
+							return
+						case <-notifier.Channel():
+							processAllPending()
 						}
 					}
-				}
-			}()
-		}
-
-		// wait long enough for all consumer to be ready for new notification.
-		consumersAllReady.Wait()
-
-		var workersAllReady sync.WaitGroup
-		workersAllReady.Add(10)
-
-		// 10 routines pushing work
-		for i := 0; i < 10; i++ {
-			go func() {
-				workersAllReady.Done()
-				for scheduledWork.Inc() <= totalWork {
-					pendingWorkQueue <- struct{}{}
-					notifier.Notify()
-				}
-			}()
-		}
+				}()
+			}
 
-		// wait long enough for all workers to be started.
-		workersAllReady.Wait()
+			// wait for all consumer to be ready for new notification.
+			synctest.Wait()
 
-		// require that all work is eventually consumed
-		if !conditionEventuallySatisfied(func() bool { return consumedWork.Load() == totalWork }, 3*time.Second, 100*time.Millisecond) {
-			require.Fail(t, "timed out", "only consumed %d units of work but expecting %d", consumedWork.Load(), totalWork)
-		}
-	}
+			// 10 routines pushing work
+			for range 10 {
+				go func() {
+					for scheduledWork.Load() <= totalWork {
+						pendingWorkQueue <- struct{}{}
+						scheduledWork.Inc()
```

Comment:
previously, `scheduledWork` was incremented `totalWork` times, plus one additional time per producer on the final failing check. moving the increment to track work actually pushed into the channel lets us verify all scheduled work was processed.

Reply:
What if we remove the atomic `scheduledWork` counter, and instead have each producer goroutine push a known number of jobs before exiting? Then we know exactly how much work was scheduled when the producers exit without active coordination via the atomic.